[#64605] Migrate to new esbuild-based Angular build system

myabc opened the pull request:
# Ticket

https://community.openproject.org/wp/64605

# What are you trying to accomplish?

This PR migrates OpenProject to the new esbuild‐based Angular build system.

https://angular.dev/tools/cli/build-system-migration

This is a prerequisite for upgrading to modern versions of Angular (we are on version 17, the latest release is 20) as well as TypeScript. Migrating to esbuild is a step that we should take before introducing new technologies to our frontend build pipeline (e,g `.tsx` and React for BlockNote support in #19004)

esbuild has a reputation for being a fast bundler. Having said that, we have yet to do any of our own benchmarking.

In detail:

- Updates tsconfig and build configuration files (`angular.json`) to support esbuild.
- Uses Vite under-the-hood as a development server.
- Modifies asset handling and dynamic imports to better align with the new build pipeline.
- Standardises third‐party library imports (e.g. moment, dragula, Mousetrap) and global vendor initialisation.

# What approach did you choose and why?

The migration has been performed in a couple steps:

1. [migration to `browser-esbuild` "compatibility" builder](https://angular.dev/tools/cli/build-system-migration#manual-migration)
2. [migration to new `application` builder](https://angular.dev/tools/cli/build-system-migration#manual-migration-to-the-new-application-builder)
3. [switching to the custom esbuild `application` builder](https://github.com/just-jeb/angular-builders/tree/master/packages/custom-esbuild) from @angular-builders

the custom esbuild builder extends the `application` builder to allow additional options to be passed (`plugins`, `indexHtmlTransformer`), potentially allowing us to take advantage of the [esbuild plugin ecosystem](https://esbuild.github.io/plugins/).

### Out-of-scope

The following improvements are worth considering, but not part of this PR/Work Package:

- Updating Angular itself. There are separate (draft) PRs:
  - #19287
  - #19288 
- Updating TypeScript - upgrading Angular v19 will also bump TS to the latest version, 5.8.3.
- Switching from our own hand-rolled asset pipeline integration (i.e. `OpenProject::Assets`) to a more standard Rails solution like [jsbundling-rails](https://github.com/rails/jsbundling-rails) + esbuild.
- Switching to a vanilla esbuild configuration (`ng` CLI used to have an `eject` command).

# Merge checklist

- [ ] Compression e.g. with Terser (e.g. https://makandracards.com/makandra/622791-esbuild-compressing-javascript-harder-terser)
- [x] Added/updated tests
  - [x] Frontend tests passing
  - [x] Capybara feature specs passing
- [ ] Added/updated documentation in Lookbook (patterns, previews, etc)
- [x] Tested major browsers (Chrome, Firefox, Edge, ...)

myabc committed:
Bump @types/dragula to 3.7.5

myabc committed:
Bump @angular-devkit/build-angular to 17.3.17

myabc committed:
[#64605] First pass migrating to browser-esbuild

* Switches ng build to browser-esbuild builder.
* Fixes static and dynamic imports.
* Switches to relative asset paths.
* Modernizes tsconfig files.
* Vendors jquery-ui. / HT https://gorails.com/episodes/how-to-use-jquery-with-esbuild
* Fixes various out-of-date path references.

See https://angular.dev/tools/cli/build-system-migration\#manual-migration-to-the-compatibility-builder

myabc committed:
Type fixes: assume catch variables are unknown

myabc committed:
Make rake assets:precompile work with esbuild

Updates `Assets.rebuild_manifest!` hash format.

myabc committed:
Defer loading of inline <script> tags with jQuery

myabc committed:
Remove unused @angular-builders/custom-webpack dependency

myabc committed:
Switch to application bundler

See: https://angular.dev/tools/cli/build-system-migration#for-existing-applications

myabc committed:
Switch to custom ESBuild builder

See: https://github.com/just-jeb/angular-builders/tree/18.x.x/packages/custom-esbuild

myabc commented on frontend/src/app/core/setup/init-locale.ts:
```diff
@@ -37,9 +37,11 @@ export function initializeLocale() {
+
+  window.I18n = i18n;
 
   moment.locale(userLocale);
```
N.B. the above line does not appear to do anything.

FWIW [moment docs](https://momentjs.com/docs/#/i18n/changing-locale/) state that:

> As of 2.8.0, changing the global locale doesn't affect existing instances.
> 
> ```js
> moment.locale('fr');
> var m = moment(1316116057189);
> m.fromNow(); // il y a une heure
> 
> moment.locale('en');
> m.fromNow(); // il y a une heure
> moment(1316116057189).fromNow(); // an hour ago
> ```

I'm not sure if this is the issue here.

myabc replied:
The two remaining failing feature specs appear to be relate to moment not loading locale date correctly:

```
rspec ./spec/features/notifications/settings/workdays_settings_spec.rb:110 # Workday notification settings with the my page behaves like workday settings with Chinese Simplified locale and start of week setting defined displays week days in Chinese (bug #49848)
rspec ./spec/features/notifications/settings/workdays_settings_spec.rb:110 # Workday notification settings with the user administration page behaves like workday settings with Chinese Simplified locale and start of week setting defined displays week days in Chinese (bug #49848)
```

@cbliard - it looks like you looked into this back in 2023 - #13607 / [op 49848](https://community.openproject.org/wp/49848).

cbliard replied:
> N.B. the above line does not appear to do anything.
> 
> FWIW [moment docs](https://momentjs.com/docs/#/i18n/changing-locale/) state that:
> 
> > As of 2.8.0, changing the global locale doesn't affect existing instances.
> > ```js
> > moment.locale('fr');
> > var m = moment(1316116057189);
> > m.fromNow(); // il y a une heure
> > 
> > moment.locale('en');
> > m.fromNow(); // il y a une heure
> > moment(1316116057189).fromNow(); // an hour ago
> > ```
> 
> I'm not sure if this is the issue here.

Yes but no.

> As of 2.8.0, changing the global locale doesn't affect **existing** instances.
> ```js
> moment.locale('fr');
> var m = moment(1316116057189);
> m.fromNow(); // il y a une heure
> 
> moment.locale('en');
> m.fromNow(); // il y a une heure
> moment(1316116057189).fromNow(); // an hour ago
> ```

`m` keeps it's locale as it was at its instanciation: fr
and a new instance, `moment(1316116057189)`, gets the global locale which is now en.

changing the global locale doesn't affect **existing** instances, or to say it another way it affects only new instances.

When doing `moment.locale('fr')`, it should try to load the locale dynamically. For now, it knows only about `de` and `en-gb` because they are hard-loaded in `frontend/src/app/core/setup/init-vendors.ts`:

```js
import moment from 'moment';
import 'moment/locale/de';
import 'moment/locale/en-gb';
```

I'm pretty sure `de` locale could be removed once the locale resources loading works properly for moment.

Because yes for now it does not work properly:

```js
moment.locale('zh-CN')
=> 'en-gb'
moment.locale('fr-FR')
=> 'en-gb'
moment.locale('de-DE')
=> 'de'
moment.locale('en-US')
=> 'en'
moment.locale('en-GB')
=> 'en-gb'
```

it knows only the ones already loaded. It was more dynamic before. It's only about moment.

To experience it, use another language than en or de (like fr), then go to personal settings > emails reminders. Here the day names are given by moment and they use en as a fallback, because it did not manage to load the locales correctly:

![image](https://github.com/user-attachments/assets/e03b7e55-512b-4152-be25-46b2a3458780)

I tried looking up documentation about how to do dynamic import with esbuild, but [it looks like it's complicated](https://github.com/angular/angular-cli/issues/26904).

Regarding the bug I fixed back then, it was simply that we called `moment.updateLocale('zh-CN', {...}) and it did not work because it knows nothing about `zh-CN`. But it knows `zh-cn`, and it can convert `zh-CN` into a known locale name when calling `moment.locale('zh-CN')`. The recognized locale can then be fetched with `moment.locale()` which would now return `zh-cn`. That's why my name appear in the git blame and why I wrote those two tests to save us from regressions:

![image](https://github.com/user-attachments/assets/ab8f051f-6c07-4552-993b-e069eba204e9)

I hope that will help finding a solution.

myabc commented on frontend/src/app/core/setup/init-vendors.ts:
```diff
@@ -31,36 +31,43 @@
+// import 'jquery-ui/ui/widgets/sortable';
+// import 'jquery-ui/ui/widgets/dialog';
+// import 'jquery-ui/ui/widgets/tooltip';
+import 'core-vendor/jquery-ui-1.14.1/jquery-ui';
```
FYI I vendored all of the above dependencies via the jQuery UI [Download Builder](https://jqueryui.com/download/)

myabc commented on frontend/src/proxy.conf.mjs:
```diff
@@ -1,15 +1,14 @@
+export default [
   {
-    "context": ['/**'],
+    "context": [
+      '/api', 
+      '/assets/frontend/media'
+    ],
```
this glob doesn't work with Vite - I'm not sure if there are any other endpoints we need to take into account?

oliverguenther replied:
I actually think we could get rid of this proxy, it was designed to work with the angular built-in serve with hot reload, but I'm rather sure all devs by now use the rails server instead.

myabc replied:
@oliverguenther I'll have to double check, but I think this proxy is still needed for the fonts to work.

myabc commented on app/views/layouts/base.html.erb:
```diff
@@ -149,33 +149,35 @@ See COPYRIGHT and LICENSE files for more details.
 <%# Properly decides main menu expanded state and width before its drawn. Fixes flickering side menu
         where menu is first expanded, then being collapsed in angular. %>
-<%= nonced_javascript_tag do %>
+<%= nonced_javascript_tag defer: true do %>
```
FYI setting the script's `defer` attribute - along with `window.addEventListener('turbo:load', () => {` is necessary to ensure jQuery has been loaded before this code is called.

Alternatively, we could investigate loading jQuery as a separate, external dependency (without the `type: "module"`):

    = javascript_include_tag "jquery", "data-turbo-track": "reload"

oliverguenther committed:
Fix import in page.controller

oliverguenther committed:
Rename to tsx, and fix import

myabc committed:
Fix TurboHelpers import: add path to tsconfig

myabc committed:
Statically import moment locales

For the time-being we import locale data for all locales for which we
have Crowdin data.

myabc committed:
Fix dynamic controller imports

myabc committed:
Fix dragula import, remove conflicting global
